core/src/main/java/org/apache/calcite/tools/RelBuilder.java

```diff
@@ -230,6 +230,18 @@ public RelDataTypeFactory getTypeFactory() {
     return cluster.getTypeFactory();
   }
 
+  /** Return the RelFactories Struct. */
+  public RelFactories.Struct getRelFactoriesStruct() {
+    return this.struct;
+  }
+
+  /** Set the RelFactories Struct. Conventions can use a different RelFactories struct to
+   *  create physical RelNode.*/
+  public RelBuilder setRelFactoriesStruct(RelFactories.Struct newStruct) {
+    this.struct = newStruct;
+    return this;
+  }
+
```

Comment:
Instead of adding `#setRelFactoriesStruct`, how abount adding `RelBuilder #copy(RelFactories.Struct newStruct) `which returns a RelBuilder but with new RelFactories?

Reply:
Then it's not really transforming a rel builder, it would be more like "creating" a new rel builder.

Reply:
Yep, it is exactly what I want. Because I think it's not a good practise to have a multable object in RelBuilder.

Reply:
We had a discussion in the JIRA about why it's not a good idea.

Reply:
Let's move the config to `RelBuilder.Config`, we do not really set or get anything on the `RelBuilder` because it's configuration should always be immutable. An immutable configuration makes the Object more deterministic and we know its state when we create it.

Reply:
Can you take a look at the JIRA discussion? If we cannot address the two points I raised, we couldn't use the way you suggest.